File: frontend/src/types/job.ts
// Mirrors backend/app/schemas/job.py -- keep these two in sync by hand for
// now (no shared codegen yet), same convention as types/skill.ts.
import type { SkillCategory } from "@/types/skill";

// Canonical home for the "at most N selected postings" rule -- mirrors
// schemas/job.py's MAX_SELECTED_POSTINGS. Feeds the full roadmap
// (types/roadmap.ts imports this rather than redefining it).
export const MAX_SELECTED_POSTINGS = 10;

// Page size for the /jobs/match result list -- mirrors schemas/job.py's
// DEFAULT_PAGE_SIZE. The backend clamps a request to schemas/job.py's
// MAX_PAGE_SIZE (50); the UI only ever asks for this.
export const DEFAULT_PAGE_SIZE = 15;

export interface MatchedJobPosting {
  id: string;
  company_name: string;
  title: string;
  location: string | null;
  url: string | null;
}

export type RequirementLevel = "required" | "preferred";

export interface SkillGapItem {
  skill_id: string;
  name: string;
  category: SkillCategory;
  requirement_level: RequirementLevel;
  user_has: boolean;
}

// See backend/app/schemas/job.py's DataFreshness/TaskStatus for the
// meaning of each value -- Phase 3 of the /jobs/match latency work
// (database-only + non-blocking + background-refresh polling).
export type DataFreshness = "fresh" | "stale" | "pending";
export type TaskStatus = "queued" | "running" | "completed" | "partial_failure" | "failed";

export interface JobMatchResponse {
  target_position: string;
  // One page (<= page_size) of the full match set -- see the pagination
  // fields below. skill_gap stays aggregated over EVERY match so the gap
  // chart doesn't change as the user pages through.
  postings: MatchedJobPosting[];
  skill_gap: SkillGapItem[];
  freshness: DataFreshness;
  // Set whenever a background refresh was enqueued (freshness is
  // "stale" or "pending") -- poll GET /jobs/match/status/{task_id}.
  task_id: string | null;
  // Pagination over the full match set. `postings` is page `page` of
  // `total_count` results; a page past `total_pages` comes back with an
  // empty `postings` and these counts still correct.
  page: number;
  page_size: number;
  total_count: number;
  total_pages: number;
}

export interface JobMatchStatusResponse {
  task_id: string;
  status: TaskStatus;
  data_freshness: DataFreshness;
  last_updated_at: string | null;
  error_summary: string | null;
}

File: backend/app/schemas/job.py
"""Job posting schemas -- response shapes for api/routes/jobs.py."""

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel

# "fresh" -- cached data is within the freshness window, returned as-is.
# "stale" -- cached data exists but is old; returned immediately anyway
# (stale-while-revalidate) alongside a task_id for the background refresh
# already enqueued. "pending" -- no data exists yet at all; postings/
# skill_gap are empty and the caller should poll task_id until it's done.
DataFreshness = Literal["fresh", "stale", "pending"]

TaskStatus = Literal["queued", "running", "completed", "partial_failure", "failed"]

# Canonical home for the "at most N selected postings" rule -- the
# User_Job_Selection a user makes on /jobs feeds POST /roadmaps
# (schemas/roadmap.py imports this instead of redefining it).
MAX_SELECTED_POSTINGS = 10

# POST /jobs/match result pagination. page_size is deliberately small --
# Step 3's selection UI shows one page at a time, and a broad target
# position ("Software Engineer") can match many hundreds of active
# postings; MAX_PAGE_SIZE bounds a client that asks for too much at once.
# The match filter itself (ingestion.query_normalization.title_matches_
# query) runs in Python, not SQL, so this paginates the response payload
# and the rendered list -- not the underlying active-postings read.
DEFAULT_PAGE_SIZE = 15
MAX_PAGE_SIZE = 50


class MatchedJobPosting(BaseModel):
    id: uuid.UUID
    company_name: str
    title: str
    location: str | None
    url: str | None


class SkillGapItem(BaseModel):
    """One skill required/preferred by at least one matched posting, and
    whether the current user already has it (per their `user_skill` rows).
    A real per-posting match score is services/matching.py's job, later --
    this is the aggregate, lightweight version.
    """

    skill_id: uuid.UUID
    name: str
    category: Literal["technical", "soft"]
    requirement_level: Literal["required", "preferred"]
    user_has: bool


class JobMatchResponse(BaseModel):
    target_position: str
    # `postings` is ONE page (at most `page_size` items) of the full match
    # set -- see the pagination fields below. `skill_gap` is still
    # aggregated across EVERY matched posting, not just this page, so the
    # gap chart doesn't lurch around as the user flips pages.
    postings: list[MatchedJobPosting]
    skill_gap: list[SkillGapItem]
    # See DataFreshness above. task_id is set whenever a background
    # refresh was enqueued (freshness in ("stale", "pending")) -- poll
    # GET /jobs/match/status/{task_id} for completion.
    freshness: DataFreshness
    task_id: uuid.UUID | None = None
    # Pagination over the full match set. `total_count`/`total_pages`
    # describe every posting matching the target position; `postings`
    # above is page `page` of that. Requesting a page past `total_pages`
    # returns an empty `postings` with these counts still correct -- a
    # valid empty page, not an error. Defaults keep older/!test callers
    # that build this without pagination working.
    page: int = 1
    page_size: int = DEFAULT_PAGE_SIZE
    total_count: int = 0
    total_pages: int = 0


class JobMatchStatusResponse(BaseModel):
    task_id: uuid.UUID
    status: TaskStatus
    data_freshness: DataFreshness
    last_updated_at: datetime | None
    # Sanitized -- never a raw exception string. None unless status is
    # 'failed'/'partial_failure'.
    error_summary: str | None = None
